### PR TITLE
Add Day1 and Day2 solutions with tests

### PR DESCRIPTION
**Pull request by ya55en**
Add Day1 and Day2 solutions with tests.

### CODE REVIEW

File: src/day1/array_ext.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#: Extend Array with scan_zzz methods (see method docs).
+
+class Array
```
Comment (emil-rb) [inline src/day1/array_ext.rb]: Although extending Array is possible, it's considered a bad practice for general code. It should be done only when the added method is used extensively (i.e. in Rails).

File: src/day1/expense_report_fix.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+require 'optparse'
+require 'ostruct'
+
+require_relative '../utils'
+require_relative './array_ext'
+
+module ExpenseReportFix
+
+  def ExpenseReportFix.scan_pairs(amounts, expected_sum = 2020)
```
Comment (emil-rb) [inline src/day1/expense_report_fix.rb]: When defining module-level methods the idiom is to do `extend self` in the beginning.
https://www.bigbinary.com/blog/extend-self-in-ruby

File: src/day2/scan_passwds.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  #: Return true if given passwd matches original shopkeeper's policies
+  #: constraints, false otherwise.
+  def ScanPasswds.match_constraints_orig?(min, max, char, passwd)
+    count = passwd.count char
+    min <= count && count <= max
+  end
+
+  #: Return true if given passwd matches Toboggan Corporate Policies constraints,
+  #: false otherwise.
+  def ScanPasswds.match_constraints_tcp?(fst_idx, sec_idx, char, passwd)
```
Comment (emil-rb) [inline src/day2/scan_passwds.rb]: Don't abbreviate variables. We consider this a bad practice.

File: src/day2/scan_passwds.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  #: Return true if given passwd matches Toboggan Corporate Policies constraints,
+  #: false otherwise.
+  def ScanPasswds.passwd_valid?(line, match_method_sym)
+    match_data = line.strip.match(/^(\d+)-(\d+)\s+(\w):\s+(.+)$/)
+    raise "Line does NOT match format: [#{line.strip}]" unless match_data
+    min_s, max_s, char, passwd = match_data.captures
+    min, max = min_s.to_i, max_s.to_i
+    # print "min=#{min}, max=#{max}, char=#{char}, passwd=#{passwd}\n"
+    match_method = ScanPasswds.method(match_method_sym)
```
Comment (emil-rb) [inline src/day2/scan_passwds.rb]: Clever! But we don't like clever code :) An `if` would've been sufficient.

File: src/day2/scan_passwds.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    match_method.call min, max, char, passwd
+  end
+
+  #: Return the number of valid password entries in `input_file` using a match
+  #: policy defined by a method with symbol `match_method_sym`.
+  def ScanPasswds.count_valid_passwords(input_file, match_method_sym)
+    count = 0
+    file_obj = input_file.respond_to?(:read) ? input_file : File.open(input_file)
+    file_obj.each_line do |line|
```
Comment (emil-rb) [inline src/day2/scan_passwds.rb]: For bonus points you can do this in a functional way with `reduce` :)

File: src/utils.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#: Common utilities for AoC solutions project
+
+module AocUtils
+
+  def AocUtils.file_to_array(filename)
+    res = []
+    File.open(filename) do |f|
+      f.each_line do |line|
```
Comment (emil-rb) [inline src/utils.rb]: You can do `f.each_line.map { |l| l.strip.to_i }`

File: src/day1/array_ext.rb (hunk 1 above)
Comment (ya55en) [reply src/day1/array_ext.rb]: Got it, thanks, fixed in 5c87212.

File: src/day1/expense_report_fix.rb (hunk 2 above)
Comment (ya55en) [reply src/day1/expense_report_fix.rb]: Thanks, `extend self` is a nice thing! (I don't think I've seen it in any of the books, btw). This one fixed in 5c87212; also reworked `ScanPasswds` and `AocUtils` the same way.

File: src/day2/scan_passwds.rb (hunk 3 above)
Comment (ya55en) [reply src/day2/scan_passwds.rb]: Assuming you mean the arguments to `match_constraints_tcp?` - fixed in df1b427.

File: src/day2/scan_passwds.rb (hunk 4 above)
Comment (ya55en) [reply src/day2/scan_passwds.rb]: Indeed. Fixed in df1b427. Also fixed command line options propagation in `ExpenseReportFix` the same way.

File: src/day2/scan_passwds.rb (hunk 5 above)
Comment (ya55en) [reply src/day2/scan_passwds.rb]: Done with 8c5eadb :)

File: src/utils.rb (hunk 6 above)
Comment (ya55en) [reply src/utils.rb]: Yep, much more elegant. Done in 6b32ef8.

File: src/day2/scan_passwds.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+require 'optparse'
+require 'ostruct'
+
+module ScanPasswds
+
+  extend self
+
+  #: Return true if given passwd matches original shopkeeper's policies
+  #: constraints, false otherwise. (See day2/problems.txt for details.)
+  def match_constraints_orig?(min, max, char, passwd)
```
Comment (emil-rb) [inline src/day2/scan_passwds.rb]: I meant to not abbreviate the names of the variables: `passwd` -> `password`. Min, max and char are fine. Those are widespread enough.

File: src/day2/scan_passwds.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    count = passwd.count char
+    min <= count && count <= max
+  end
+
+  #: Return true if given passwd matches Toboggan Corporate Policies (TCP)
+  #: constraints, false otherwise. (Note that TCP policy authors count from 1.
+  #: See day2/problems.txt for details.)
+  def match_constraints_tcp?(first, second, char, passwd)
+    (passwd[first - 1] == char) ^ (passwd[second - 1] == char)
```
Comment (emil-rb) [inline src/day2/scan_passwds.rb]: What does this `^` do in this case?

File: tests/day1-tests/test_expense_report_fix.rb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    def test_scan_pairs__size_1
+      res = []
+      ExpenseReportFix.scan_pair_combinations(1) { |i, j| res << [i, j] }
+      assert_equal [], res
+    end
+
+    def test_scan_triads__size_4
+      res = []
+      ExpenseReportFix.scan_triad_combinations(4) { |i, j, k| res << [i, j, k] }
```
Comment (emil-rb) [inline tests/day1-tests/test_expense_report_fix.rb]: You can fix this need to pass a block by doing this: https://stackoverflow.com/a/66059990/135665

File: src/day2/scan_passwds.rb (hunk 7 above)
Comment (ya55en) [reply src/day2/scan_passwds.rb]: I see. fixed in 44e54b6.

File: tests/day1-tests/test_expense_report_fix.rb (hunk 9 above)
Comment (ya55en) [reply tests/day1-tests/test_expense_report_fix.rb]: Yep, nice, thank you! Done in e07ebc5.

File: src/day2/scan_passwds.rb (hunk 8 above)
Comment (ya55en) [reply src/day2/scan_passwds.rb]: Xor. This is an implementation of the "Toboggan Corporate Policy" for authentication, namely:
> Exactly one of these positions must contain the given letter.

Thus `match_constraints_tcp?` returns `true` iff exactly one of the positions contains the required character.

Comment (emil-rb) [reply src/day2/scan_passwds.rb]: OK, cool. Now I learned this operator can be used for booleans as well as bit math.

**Review by emil-rb — APPROVED** [review]

**Pull request merged**